### ISSUE TITLE
Add callbacks to Main::setup / Main::start / Main::cleanup, so to allow much freedom to modules.

### ISSUE THREAD

**Issue by AndreaCatania** [godotengine/godot-proposals#1593]
**Describe the project you are working on:**
I'm working on a game, and I'm creating various C++ modules. Couple of those modules, need to know when the engine setup is over, and so the `SceneTree` or other singletons are ready to be used.

**Describe the problem or limitation you are having in your project:**
Those modules try to connect to some events like `process` and so on, but since the engine is not ready, there is no way to connects to those events. Right now, from the `register_MODULE_types()` it's not possible to know when the engine starts.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
The idea is to add 3 callbacks to the `Main` class:
- on_setup_done
- on_start
- on_cleanup

From the `register_MODULE_types()` will be possible to add callbacks in the following way:
```cpp
void register_MODULE_types(){
    Main::add_callback(&my_setup_function, Main::SETUP_CALLBACK);
}
```
All the registered callbacks will be called when the setup/start terminates or the cleanup is started.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
Right before the `Main::setup2` is done https://github.com/godotengine/godot/blob/master/main/main.cpp#L1719-L1720 the following code is executed:
```cpp
for(int i = 0; i < on_setup_done_callbacks.size(); i+=1){
    on_setup_done_callbacks[i]();
}
```
So the modules, will be able to interact with the engine extensively.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
Low level feature.

**Is there a reason why this should be core and not an add-on in the asset library?:**
Low level feature.

**Comment by Xrayez** [comment]
This is needed because of issues like Zylann/godot_voxel#189.

Godot 4.0 adds ability to [`preregister_types()`](https://docs.godotengine.org/en/latest/development/cpp/custom_modules_in_cpp.html#customizing-module-types-initialization), but it has no corresponding `preunregister_types()`. Of course having `Main` callbacks as proposed here could also provide an extensive solution, I believe. 🙂

**Comment by Zylann** [comment]
Would this also allow my module to register a "process" callback without needing a node in the tree? I have some sort of server which needs to run an update loop like some other servers do (Physics Server, Visual Server) but I could not find a way to insert this within the main loop.

**Comment by AndreaCatania** [comment]
An unpolished way of doing it is via message queue, but it's just a hack.
Instead, with #1592, you can obtain the SceneTree singleton (into the setup_done callback) and register the function to the process event. In this way your singleton is naturally called each frame or physics frame.

**Comment by AndreaCatania** [comment]
Yes, you can obtain the SceneTree singleton (into the setup_done callback) and register the function to the process event. In this way your singleton is naturally called each frame or physics frame without the need to use a node.

**Comment by Xrayez** [comment]
I've noticed there's `SceneTree::add_idle_callback()` which can be used in modules, but this is only useful for flushing custom deferred calls since that's how `MessageQueue` flushing is currently used throughout the engine (those idle callbacks are called on `_call_idle_callbacks()`), you can actually see it in action in goostengine/goost#51.

I suspect that it might not be always thread-safe, so that's where we'd also like to have ability to register process callbacks as well. Due to this, I think that it may be also required to check whether you're inside physics frame with `Engine.is_in_physics_frame()` too if you do use `SceneTree::add_idle_callback()`, but I'm not 100% sure about it.

> Yes, you can obtain the SceneTree singleton (into the setup_done callback) and register the function to the process event.

@AndreaCatania not sure whether I understand this correctly, do you imply that there's some other/existing mechanism already available for registering process callbacks within the engine (just like `SceneTree::add_idle_callback()`, or perhaps we're talking about the same thing already)?

**Comment by AndreaCatania** [comment]
I've implemented this feature here: https://github.com/godotengine/godot/pull/50179 @Xrayez check it, maybe it's cleaner what I meant by checking the code.
